homework-submission/Week2/App.js

```diff
@@ -0,0 +1,119 @@
+// creates a todo item as a <li> element with props which consists description, deadline and className
+const TodoItem = props => {
+  const { description, deadline, className, deleteItem } = props;
+  return (
+    <li className={className} onClick={deleteItem}>
+      {description}, {deadline}
+    </li>
+  );
+};
+
+class App extends React.Component {
+  state = {
+    todoList: [
+      {
+        id: 1,
+        description: 'Get out of bed',
+        deadline: '2017-09-11',
+        done: true,
+      },
+      {
+        id: 2,
+        description: 'Brush teeth',
+        deadline: '2017-09-10',
+        done: false,
+      },
+      {
+        id: 3,
+        description: 'Eat breakfast',
+        deadline: '2017-09-09',
+        done: false,
+      },
+    ],
+
+    newTodo: {
+      id: '',
+      description: '',
+      deadline: '',
+      done: false,
+    },
+  };
+
+  AddNewTodo = event => {
+    event.preventDefault();
+    const newTodo = {
+      id: Date.now(),
+      description: this.state.newTodo.description,
+      deadline: this.state.newTodo.deadline,
+      done: false,
+    };
+    this.setState({ todoList: [...this.state.todoList, newTodo] });
+  };
+
+  DeleteItem = id => {
+    const { todoList } = this.state;
+    const indexOfItem = todoList.findIndex(elem => elem.id === id);
+    todoList.splice(indexOfItem, 1);
```

Comment:
if you tried the following you will notice a direct change of the state
```javascript
    console.log(this.state.todoList.length) // if this 3
    todoList.splice(indexOfItem, 1); // you are splicing here!
    console.log(this.state.todoList.length) // this is 2 and it should stay 3
```

This direct change shouldn't be happening, the main reason of `.setState` is making things asynchronous.

the `.setState` method isn't only for redering the content, it's made on top of a complete life cycle happening on behalf of your program, and we want to make sure we are not changing the state by hand as here.

the best solution here would be either to make a deep copy of the array, or making a new array, from js2 we remember having `.filter` will return a new array, and this is the best example of it, and where it should go.

Reply:
I used .filter  as you mentioned.